pkg/kubelet/network/kubenet/kubenet_linux.go

```diff
@@ -0,0 +1,301 @@
+// +build linux
+
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubenet
+
+import (
+	"fmt"
+	"net"
+	"strings"
+	"syscall"
+
+	"github.com/vishvananda/netlink"
+
+	"github.com/appc/cni/libcni"
+	"github.com/golang/glog"
+	"k8s.io/kubernetes/pkg/api"
+	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/pkg/kubelet/dockertools"
+	"k8s.io/kubernetes/pkg/kubelet/network"
+	"k8s.io/kubernetes/pkg/util/bandwidth"
+)
+
+const (
+	KubenetPluginName    = "kubenet"
+	BridgeName           = "cbr0"
+	DefaultCNIDir        = "/opt/cni/bin"
+	DefaultInterfaceName = "eth0"
+)
+
+type kubenetNetworkPlugin struct {
+	host      network.Host
+	netConfig *libcni.NetworkConfig
+	cniConfig *libcni.CNIConfig
+	shaper    bandwidth.BandwidthShaper
+
+	podCIDRs map[kubecontainer.DockerID]string
+	MTU      int
+}
+
+func NewPlugin() network.NetworkPlugin {
+	return &kubenetNetworkPlugin{
+		podCIDRs: make(map[kubecontainer.DockerID]string),
+		MTU:      1460,
+	}
+}
+
+func (plugin *kubenetNetworkPlugin) Init(host network.Host, nodeAddrs []api.NodeAddress) error {
+	plugin.host = host
+	plugin.cniConfig = &libcni.CNIConfig{
+		Path: []string{DefaultCNIDir},
+	}
+
+	// Grab MTU of interface that has the node address
+	link, err := getLinkForNodeAddress(nodeAddrs)
+	if err == nil {
+		plugin.MTU = link.Attrs().MTU
+		glog.V(5).Infof("Using host interface %s MTU %d as bridge MTU", link.Attrs().Name, plugin.MTU)
+	} else {
+		glog.Warningf("%s", err)
+	}
+
+	return nil
+}
+
+func getLinkForNodeAddress(nodeAddrs []api.NodeAddress) (netlink.Link, error) {
+	// Grab MTU of interface that has the node address
+	links, err := netlink.LinkList()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to list host links: %v", err)
+	}
+
+	nodeAddr := ""
+	for _, a := range nodeAddrs {
```

Comment:
I've been thinking about this.  It seems like sort of a hard generic problem, at first, but maybe the answer is actually simpler.

Shouldn't the MTU of the bridge and veths just be the smallest MTU of any "real" interface on the system?

Could we simply iterate all "up" interfaces, filter out loopbacks, and take the min MTU?  That seems equivalent to choosing the default route for simple configs and more likely to succeed for complex configs.

What do you think?  Otherwise just hardcode to 1460 and leave a TODO to detect a better value.  It's not worse than we have today.  All the nodeAddrs plumbing can die, I think.